### ISSUE TITLE
discrepancy of the FSDP training step with the **alpaca_dataset** dataset

### ISSUE THREAD

**Issue by mathmax12** [meta-llama/llama-recipes#641]
### 🚀 The feature, motivation and pitch

I’m new to working with FSDP and testing the FSDP fine-tuning with **alpaca_dataset** dataset. I have a single node with 8 GPUs and have set the batch size to 16. The dataset I’m training on contains 49,402 samples ([Alpaca-cleaned on Hugging Face](https://huggingface.co/datasets/yahma/alpaca-cleaned)).
![image](https://github.com/user-attachments/assets/51c961eb-724e-4a6a-9ae9-3be29320cb81)

Based on my understanding, when using FULL_SHARD, each GPU should process a batch of 16 samples during the forward pass, which means a total of 8x16 samples per step. Given this, I expected there to be 49402 / (8x16) ≈ 386 steps per epoch. However, the logs show only 9 steps per epoch.
![image](https://github.com/user-attachments/assets/7f76f5b4-78b8-4dae-b197-e92a85f8a0c1)

Could someone help me understand why this discrepancy is occurring?

Thanks

**Comment by mreso** [comment]
Hi @mathmax12 
Did you change the [batching strategy](https://github.com/meta-llama/llama-recipes/blob/b5f64c0b69d7ff85ec186d964c6c557d55025969/src/llama_recipes/configs/training.py#L15) from packing to padding? If you do that you should see about 386 steps. Packing will put multiple samples into a single row of the input batch to avoid padding. Another cause could be that you enabled gradient accumulation, as the shown steps are the global steps after which finish after accumulating enough gradients.

**Comment by mathmax12** [comment]
@mreso  
Thanks for your prompt reply. You are correct—I observed the number change after applying padding.

Regarding packing, it concatenates the samples to fit the model's context length. Does this mean that the packed samples need to be related to one another? If the samples are unrelated, I'm uncertain whether this approach would help the model understand any relationships, as there's a possibility that each sample within the package is independent. Does that mean this strategy is only for the dataset that have the samples with highl correlation? 

For the padding strategy, if each sample is very short, will the padding extend it to the model's full context length before processing

**Comment by mreso** [comment]
For padding he longest sample in a batch dictates the amount of padding. So its only up to the longest sequence, not context length.

Packing is a strategy also used during pre-training. Samples are encapsulated within beginning of sequence and end of sequence token. The model should learn to cut off context before and after. It is possible to create a mask which masks the different sequences from each other but we have not implemented this change yet.
